### PR TITLE
fix: handle invalid NSUnderlyingErrorKey

### PR DESCRIPTION
**Pull request by jmkrein**
## :scroll: Description

Fix an app crash that occurs when trying to log an invalid value for the key `NSUnderlyingErrorKey` within the `userInfo` of an `NSError`.

## :bulb: Motivation and Context

Fixes https://github.com/getsentry/sentry-cocoa/issues/4143

Huge credit to [dohyeondk](https://github.com/dohyeondk) for helping identify the root cause of this issue. 

## :green_heart: How did you test it?

Added two new unit tests.

## :pencil: Checklist

You have to check all boxes before merging:

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [x] I updated the docs if needed.
- [ ] Review from the native team if needed.
- [x] No breaking change or entry added to the changelog.
- [x] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

### CODE REVIEW

**Comment by krystofwoldrich** [comment]
Hi @jmkrein,
thank you for opening this fix,
we will review it shortly, excuse the slower response time due to ongoing PTOs/Holidays.

**Review by krystofwoldrich — APPROVED** [review]
The fix looks good to me! But before we can merge it, one of the code owner will need to review it as well.

**Comment by jmkrein** [comment]
No worries on the delay, I totally understand. Thank you for the review, @krystofwoldrich!

**Comment by brustolin** [comment]
We just need an entry for the CHANGELOG

File: Sources/Sentry/SentryClient.m
Comment (armcknight) [inline Sources/Sentry/SentryClient.m:263]: We should consider logging and/or client reporting for invalid underlying errors.

Comment (jmkrein) [reply Sources/Sentry/SentryClient.m:273]: Good call. `SENTRY_LOG_*` looked like the most common logging pattern so I continued with that.

**Pull request merged**
